### PR TITLE
fix Issue 15989 - Initializing manifest constants with CTFE allocated…

### PR DESCRIPTION
**Pull request by WalterBright**
… data

### CODE REVIEW

**Comment by DmitryOlshansky** [comment]
This still doesn't answer the question of why did it work in Linux but not on Windows

**Review by UplinkCoder — APPROVED** [review]
LGTM

**Comment by UplinkCoder** [comment]
@DmitryOlshansky It does not work on either. 
Because the pointer of the object is invalid at runtime,
The virtual opCall will call into a random memory location.
depending on what data is where the vtbl is expected.
Sometimes it does not produce a segfault on windows either.

**Comment by WalterBright** [comment]
The existing compiler explicitly disallowed CTFE class objects being used to initialized enums. But the compiler neglected to check aggregate initializers for those objects. This PR fixes that, and that's all it does.

It's possible a future compiler could make those work, but that is beyond the scope of this PR.

File: src/declaration.d [hunk 1]
```diff
@@ -1949,39 +1949,44 @@ extern (C++) class VarDeclaration : Declaration
...
+                    else if (e.op == TOKstructliteral)
+                        return arrayHasInvalidEnumInitializer((cast(StructLiteralExp)e).elements);
+                    else if (e.op == TOKassocarrayliteral)
                     {
-                        error(": Unable to initialize enum with class or pointer to struct. Use static const variable instead.");
+                        AssocArrayLiteralExp ae = cast(AssocArrayLiteralExp)e;
+                        return arrayHasInvalidEnumInitializer(ae.values) ||
+                               arrayHasInvalidEnumInitializer(ae.keys);
                     }
+                    return false;
```
Comment (andralex) [inline src/declaration.d]: "When all branches of an if/else chain return, switch to a simpler pattern that removes all `else`s thus eliminating the redundant control flow and making the code easier to follow by reviewers, readers, and maintainers".

```d
                    if (e.op == TOKclassreference)
                        return true;
                    if (e.op == TOKaddress && (cast(AddrExp)e).e1.op == TOKstructliteral)
                        return true;
                    if (e.op == TOKarrayliteral)
                        return arrayHasInvalidEnumInitializer((cast(ArrayLiteralExp)e).elements);
                    if (e.op == TOKstructliteral)
                        return arrayHasInvalidEnumInitializer((cast(StructLiteralExp)e).elements);
                    if (e.op == TOKassocarrayliteral)
                    {
                        AssocArrayLiteralExp ae = cast(AssocArrayLiteralExp)e;
                        return arrayHasInvalidEnumInitializer(ae.values) ||
                               arrayHasInvalidEnumInitializer(ae.keys);
                    }
                    return false;
```

File: src/declaration.d [hunk 2]
```diff
@@ -1949,39 +1949,44 @@ extern (C++) class VarDeclaration : Declaration
             _init = _init.semantic(sc, type, sc.intypeof == 1 ? INITnointerpret : INITinterpret);
             inuse--;
         }
-        if (storage_class & STCmanifest)
+        if (_init && storage_class & STCmanifest)
         {
-            version (none)
```
Comment (andralex) [inline src/declaration.d]: yay

**Review by andralex — CHANGES_REQUESTED** [review]

File: src/declaration.d (hunk 1 above)
Comment (WalterBright) [inline src/declaration.d]: I don't understand what you're asking for.

Comment (andralex) [inline src/declaration.d]: Your code has several redundant `else`s. I'm suggesting their removal.

Comment (WalterBright) [inline src/declaration.d]: done

Comment (andralex) [inline src/declaration.d]: gracias

**Review by andralex — APPROVED** [review]

**Comment by andralex** [comment]
Auto-merge toggled on

**Pull request merged**